loadtest/user/userentity/websocket.go

```diff
@@ -98,6 +99,32 @@ func (ue *UserEntity) handlePostEvent(ev *model.WebSocketEvent) error {
 // sync with the server. Any response to the event should be made by handling
 // the same event at the upper layer (controller).
 func (ue *UserEntity) wsEventHandler(ev *model.WebSocketEvent) error {
+	if ev.EventType() == model.WebsocketEventHello {
+		if connID, ok := ev.GetData()["connection_id"].(string); ok {
+			// If we already have a connectionId present, and server sends a different one,
+			// that means it's either a long timeout, or server restart, or sequence number is not found.
+			// Then we reset sequence number to 0.
+			if ue.wsConnID != "" && ue.wsConnID != connID {
+				mlog.Debug("Long timeout, or server restart, or sequence number not found")
+				// In future, we can add the missed event callback here.
+				ue.wsServerSeq = 0
+			}
+			ue.wsConnID = connID
+		}
+	}
+
+	// Now we check for sequence number, and if it does not match,
+	// we just disconnect and reconnect.
+	if ev.GetSequence() != ue.wsServerSeq {
+		mlog.Warn("Missed websocket event", mlog.Int64("got", ev.GetSequence()), mlog.Int64("expected", ue.wsServerSeq))
+		if err := ue.Disconnect(); err != nil {
+			return err
+		}
+		// TODO: reconnect?
```

Comment:
Need some guidance here. It feels a bit cumbersome to disconnect/reconnect from the user level because all the controlling happens at the controller level. Maybe we can return a custom error to the controller and try to reconnect, but even that feels complicated. The question is - should we try to handle this case?

Reply:
I'd be opting for reconnecting without the controller even noticing. Later on we could implement a reconnect callback or some other system to notify the caller but that's not currently required. I am thinking it would be sufficient for `wsEventHandler()` to return a special error and then upon receiving that in `listen()` we reinitialize the client. We can probably even factor out and reuse the existing code which already handles reconnecting:

https://github.com/mattermost/mattermost-load-test-ng/blob/be76149215d0d018c5e2ec5814c74d49ef1db937/loadtest/user/userentity/websocket.go#L117-L132

Reply:
Thanks, will take a look.

Reply:
PTAL.

Reply:
Labels :open_mouth:

Reply:
Lol yeah, there are 2 nested loops here. I can work on an improvement PR after this if you want.

Reply:
It looks clean enough actually. Just need to test it out a bit but code changes look good.